src/MultiSigSigner.sol

```diff
@@ -0,0 +1,179 @@
+
+    struct Config {
+        uint256 threshold;
+        bytes32[] ownerKeyHashes;
+    }
+
+    /// @dev A config is mapped to a tuple of (address, keyhash)
+    /// This allows a single account, to register multiple multi-sig configs.
+    mapping(address => mapping(bytes32 => Config)) internal _configs;
+
+    function getConfig(address account, bytes32 keyHash)
+        public
+        view
+        returns (uint256 threshold, bytes32[] memory ownerKeyHashes)
+    {
+        Config memory config = _configs[account][keyHash];
+        return (config.threshold, config.ownerKeyHashes);
+    }
+
+    ////////////////////////////////////////////////////////////////////////
+    // Config Functions
+    ////////////////////////////////////////////////////////////////////////
+
+    function _checkKeyHash(bytes32 expectedKeyHash) internal view {
+        bytes32 keyHash = IDelegation(msg.sender).getContextKeyHash();
+        if (keyHash != expectedKeyHash) revert InvalidKeyHash();
+    }
+
+    function setConfig(bytes32 keyHash, uint256 threshold, bytes32[] memory ownerKeyHashes)
+        public
+    {
+        // Threshold can't be zero
+        if (threshold == 0) revert InvalidThreshold();
+
+        Config storage config = _configs[msg.sender][keyHash];
+
+        if (config.threshold > 0) {
+            revert ConfigAlreadySet();
+        }
+
+        _configs[msg.sender][keyHash] =
+            Config({threshold: threshold, ownerKeyHashes: ownerKeyHashes});
+    }
+
+    function addOwner(bytes32 keyHash, bytes32 ownerKeyHash) public {
+        _checkKeyHash(keyHash);
+
+        Config storage config = _configs[msg.sender][keyHash];
+        config.ownerKeyHashes.push(ownerKeyHash);
+    }
+
+    function removeOwner(bytes32 keyHash, bytes32 ownerKeyHash) public {
+        _checkKeyHash(keyHash);
+
+        Config storage config = _configs[msg.sender][keyHash];
+        bytes32[] storage ownerKeyHashes_ = config.ownerKeyHashes;
+        uint256 ownerKeyCount = ownerKeyHashes_.length;
+
+        bool found = false;
+
+        for (uint256 i = 0; i < ownerKeyCount; i++) {
+            if (ownerKeyHashes_[i] == ownerKeyHash) {
+                // Replace the owner to remove with the last owner
+                ownerKeyHashes_[i] = ownerKeyHashes_[ownerKeyCount - 1];
+                // Remove the last element
+                ownerKeyHashes_.pop();
+                found = true;
+                break;
+            }
+        }
+
+        if (!found) revert OwnerNotFound();
+        if (ownerKeyCount - 1 < config.threshold) revert InvalidThreshold();
+    }
+
+    function setThreshold(bytes32 keyHash, uint256 threshold) public {
+        _checkKeyHash(keyHash);
+
+        Config storage config = _configs[msg.sender][keyHash];
+
+        if (threshold == 0 || threshold > config.ownerKeyHashes.length) revert InvalidThreshold();
+
+        config.threshold = threshold;
+    }
```

Comment:
Note to self:

The `keyhash` on `msg.sender` which created the config is authorized to change the threshold.